guides/typescript.md

```diff
@@ -0,0 +1,175 @@
+# Writing and Migrating To TypeScript
+
+## Style Guidance
+
+In general, the [ts.dev](https://ts.dev/style/) style guide shall be adhered to as close as possible, with the caveat that end-of-line semiocolons are omitted unless necessary to prevent syntax errors.
+
+#### Type Definitions
+
+Prefer `interface` to `type` when possible. Interfaces are never inlined, which solves some potential issues with generated `.d.ts` files.
+
+When to use `type`:
+* The type being declared must be a Union type
+* Declaration merging may happen when it should not
+* Complicated types using utility types
+
+#### Object Literals
+When working with a structural implementation, always declare the type.
+
+Correct:
+```typescript
+interface Foo {
+  bar: string
+}
+
+const foo: Foo = {
+  bar: 'baz',
+}
+```
+Incorrect:
+```typescript
+const foo = {
+  bar: 'baz',
+}
+```
+
+#### Function Definitions
+
+Always declare types for parameters and return values. Declaring parameter types ensures that callers are providing the correct parameters. Declaring the return type ensures that the function body is returning what is intended. 
+
+_Exception_: inlined single-line arrow functions can usually have their type declarations inferred. This is especially the case if the ultimate return value is declared. Example:
+```ts
+const odds: string[] = [1,2,3,4,5]
+  .filter(n => n % 2)
+  .map(n => String(n))
+```
+
+**Prefer** inline types. Parameter types and return types often do not need to be referenced outside of the function itself. Bag parameters (object literals with more than two properties) should be declared as a named interface. If possible, do not export this interface. If a dependent needs to reference the bag type because it is composing an object literal in preparation to call the function in question, that dependent should use `Parameters<>` instead. This way the interface name can change without necessitating a change to the dependent. Use this technique, example below, as a last resort. Prefer composing the parameter bag inline, rather than assigning it to an identifier.
```

Comment:
is it worth linking out to the [parameters](https://www.typescriptlang.org/docs/handbook/utility-types.html#parameterstype) utility type?